File: python/binary-tree-general/binary_tree_max.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self) -> None:
        self.max_sum = float('-inf')

    def maxPathSum(self, root:  Optional[TreeNode]) -> int:
        self.calculateMaxPath(root)

        return self.max_sum

    def calculateMaxPath(self, node: Optional[TreeNode]) -> int:
        if not node:
            return 0

        left_gain = max(self.calculateMaxPath(node.left), 0)
        right_gain = max(self.calculateMaxPath(node.right), 0)

        current_path_sum = node.val + left_gain + right_gain

        self.max_sum = max(self.max_sum, current_path_sum)

        return node.val + max(left_gain, right_gain)

File: golang/binary-tree-general/binary_tree_max.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxPathSum(root *TreeNode) int {
	maxSum := math.MinInt32
	calculateMaxPath(root, &maxSum)

	return maxSum
}

func calculateMaxPath(node *TreeNode, maxSum *int) int {
	if node == nil {
		return 0
	}

	left_gain := max(calculateMaxPath(node.Left, maxSum), 0)
	right_gain := max(calculateMaxPath(node.Right, maxSum), 0)

	current_path_sum := node.Val + left_gain + right_gain

	*maxSum = max(*maxSum, current_path_sum)

	return node.Val + max(left_gain, right_gain)
}
